src/api/json_tokenize.cpp

```diff
@@ -0,0 +1,116 @@
+#include "json_tokenize.hpp"
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <iostream>
+#include <migraphx/errors.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+siter colon(siter start, siter end)
+{
+    return std::find_if(start, end, [](auto c) { return c == ':'; });
+}
+
+std::pair<siter, siter> key(siter start, siter end)
+{
+    // find key end
+    --end;
+    while(start != end)
+    {
+        if(*end != ' ')
+        {
+            break;
+        }
+        --end;
+    }
+    auto ke = end;
+
+    if(start == end)
+    {
+        if(*ke != '\"')
+        {
+            return {ke, ke};
+        }
+        else
+        {
+            MIGRAPHX_THROW("KEY: single quote cannot be a key!");
+        }
+    }
+
+    // find key start
+    --end;
+    while(start != end)
+    {
+        // match
+        if(*ke == '\"' and *end == '\"')
+        {
+            break;
+        }
+        else if(*ke != '\"' and (std::ispunct(*end) or *end == ' '))
+        {
+            ++end;
+            break;
+        }
+        --end;
+    }
+
+    if(start == end)
+    {
+        if(std::ispunct(*end) or *end == ' ')
+            ++end;
+    }
+
+    auto ks = end;
+
+    return {ks, ke};
+}
+
+std::string json_tokenize(const std::string& s)
+{
+    siter start = s.begin();
+    siter end   = s.end();
+    std::vector<token> tokens;
+
+    while(start != end)
+    {
+        auto colon_iter = colon(start, end);
```

Comment:
Wont this find a colon in a quote?

Reply:
I just assume our usage does not have this scenario. Do we need to consider it? Also the case of single quote inside a quote?